src/dodal/devices/oav/pin_tip_detection/oav_with_pin_tip_detection.py

```diff
@@ -0,0 +1,214 @@
+            else:
+                # Something else?
+                raise ValueError(
+                    "Unexpected data array size: expected {} (grayscale data) or {} (rgb data), got {}",
+                    num_pixels,
+                    num_pixels * 3,
+                    value_len,
+                )
+
+            start_time = time.time()
+            location = sample_detection.processArray(value)
+            end_time = time.time()
+            LOGGER.debug(
+                "Sample location detection took {}ms".format(
+                    (end_time - start_time) * 1000.0
+                )
+            )
+            tip_x = location.tip_x
+            tip_y = location.tip_y
+        except Exception as e:
+            LOGGER.error("Failed to detect pin-tip location due to exception: ", e)
+            tip_x = None
+            tip_y = None
+
+        return (tip_x, tip_y), timestamp
+
+    async def connect(self, sim: bool = False):
+        await super().connect(sim)
+
+        # Set defaults for soft parameters
+        await self.timeout.set(10.0)
+        await self.canny_upper.set(100)
+        await self.canny_lower.set(50)
+        await self.close_iterations.set(5)
+        await self.close_ksize.set(5)
+        await self.scan_direction.set(ScanDirections.FORWARD.value)
+        await self.min_tip_height.set(5)
+        await self.preprocess.set(10)  # Identity function
+        await self.preprocess_iterations.set(5)
+        await self.preprocess_ksize.set(5)
+
+    async def read(self) -> dict[str, Reading]:
+        tip_pos, timestamp = await asyncio.wait_for(
+            self._get_tip_position(), timeout=await self.timeout.get_value()
+        )
+
+        return OrderedDict(
+            [
+                (self._name, {"value": tip_pos, "timestamp": timestamp}),
+            ]
+        )
+
+    async def describe(self) -> dict[str, Descriptor]:
+        return OrderedDict(
+            [
+                (
+                    self._name,
+                    {
+                        "source": "pva://{}PVA:ARRAY".format(self._prefix),
+                        "dtype": "number",
+                        "shape": [2],  # Tuple of (x, y) tip position
+                    },
+                )
+            ],
+        )
+
+
+if __name__ == "__main__":
+    x = PinTipDetection(prefix="BL03I-DI-OAV-01:", name="edgeDetect")
+
+    async def acquire():
+        await x.connect()
+        img = await x.array_data.read()
+        tip = await x.read()
+        return img, tip
+
+    img, tip = asyncio.get_event_loop().run_until_complete(
+        asyncio.wait_for(acquire(), timeout=10)
+    )
+    print(tip)
+    print("Tip: {}".format(tip["edgeDetect"]["value"]))
+
+    try:
+        import matplotlib.pyplot as plt
+
+        plt.imshow(img[""]["value"].reshape(768, 1024, 3))
+        plt.show()
+    except ImportError:
+        print("matplotlib not available; cannot show acquired image")
```

Comment:
It's mostly a convenience so that we can run this code quickly on a beamline without needing all the rest of the bluesky infrastructure running. I'll see if I can put it somewhere better more out of the way, but I would like to keep it at least in the short term.